modules/solid_mechanics/src/neml2/userobjects/ExecuteNEML2Model.C

```diff
@@ -0,0 +1,354 @@
+    // check if the model actually has an input corresponding to this UO
+    if (!model().input_axis().has_variable(uo_var))
+    {
+      paramError("gather_uos",
+                 "The supplied user object `",
+                 uo_name,
+                 "` connects to a model input that does not exist (`",
+                 uo_var,
+                 "`).");
+    }
+    _provided_inputs.insert(uo_var);
+  }
+
+  // iterate over set of required inputs and error out if we find one that is not provided
+  std::set<neml2::VariableName> _required_inputs = model().input_axis().variable_accessors(true);
+  for (const auto & input : _required_inputs)
+    if (!_provided_inputs.count(input) && input.start_with("forces"))
+      paramWarning(
+          "gather_uos", "No user object is providing the required model input `", input, "`");
+}
+
+void
+ExecuteNEML2Model::timestepSetup()
+{
+  if (_t_step > 0)
+    advanceStep();
+}
+
+bool
+ExecuteNEML2Model::shouldCompute()
+{
+  // NEML2 computes residual and Jacobian together at EXEC_LINEAR
+  // There is no work to be done at EXEC_NONLINEAR **UNLESS** we are computing the Jacobian for
+  // automatic scaling.
+  if (_fe_problem.computingScalingJacobian())
+    return true;
+
+  if (_fe_problem.currentlyComputingResidualAndJacobian())
+    return true;
+
+  if (_fe_problem.currentlyComputingJacobian())
+    return false;
+
+  return true;
+}
+
+void
+ExecuteNEML2Model::initialize()
+{
+  _elem_to_batch_index.clear();
+  _elem_to_batch_index_cache = {libMesh::invalid_uint, 0};
+  _batch_index = 0;
+
+  if (shouldCompute())
+    _output_ready = false;
+}
+
+void
+ExecuteNEML2Model::execute()
+{
+  _elem_to_batch_index[_current_elem->id()] = _batch_index;
+  _batch_index += _qrule->n_points();
+}
+
+void
+ExecuteNEML2Model::finalize()
+{
+  if (!shouldCompute())
+    return;
+
+  // check if all gather uos collected the same amount of data, an that is is consistent with the
+  // element id to batch index map
+  for (const auto uo : _gather_uos)
+    if (uo->size() != _batch_index)
+      mooseError("Inconsistent gathered data size in ", uo->name());
+
+  try
+  {
+    initModel(_batch_index);
+
+    // Allocate the input and output once
+    if (!_in_out_allocated)
+    {
+      _in = neml2::LabeledVector::zeros(_batch_index, {&model().input_axis()});
+      _out = neml2::LabeledVector::zeros(_batch_index, {&model().output_axis()});
+      _in_out_allocated = true;
+    }
+
+    // Steps before stress update
+    // preCompute();
+
+    updateForces();
+
+    if (_t_step > 0)
+    {
+      applyPredictor();
+
+      solve();
+
+      // update output views
+      for (auto & pair : _outputs)
+        pair.second = _out(pair.first);
+
+      // output retrieved derivatives
+      for (auto & [out, in, batch_tensor_ptr] : _retrieved_derivatives)
+        *batch_tensor_ptr = _dout_din(out, in);
+
+      // Additional calculations after stress update
+      // postCompute();
+      _output_ready = true;
```

Comment:
`_output_ready` can also be true at the 0th time step, so maybe move this statement outside the if statement.

Reply:
No, that leads to a segfault. (no output is prepared)

Reply:
The execute_on for this UO is 'INITIAL LINEAR NONLINEAR', so I guess the question is why the output wasn't allocated at INITIAL.

Maybe `shouldCompute()` killed that?

Reply:
Yeah, could be, maybe the residual and jacobian flags are both false. Eitehr way `if (_t_step > 0)` certainly kills it.

Reply:
I'll revisit the execute on INITIAL issue in a follow-on PR